nextcore/endpoint/interactions.py

```diff
@@ -0,0 +1,158 @@
+        logger.debug("Incoming interaction endpoint request.")
+        try:
+            body = (await request.read()).decode("utf-8")
+            if self._verify_from_headers(request.headers, body):
+                logger.debug("Verified Discord signature headers, continuing.")
+                interaction: InteractionData = await request.json()
+                if interaction.get("type") == 1:
+                    # logger.debug("Dispatching ping.")
+                    await self.event_dispatcher.dispatch("ping", request, interaction)
+                else:
+                    await self.event_dispatcher.dispatch("interaction", request, interaction)
+
+                return None
+            else:
+                logger.debug("Failed Discord signature headers, returning 401.")
+                return web.Response(body="invalid request signature", status=401)
+
+        except JSONDecodeError:
+            logger.debug("Invalid json given, ignoring: %s", (await request.read()).decode())
+            return web.Response(status=400)
+
+    async def _handle_ping(self, request: web.Request, interaction):
+        response = web.json_response(data={"type": 1})
+        await response.prepare(request)
+        await response.write_eof()
+
+    def middleware(self, public_key: str, route: str = "/interactions"):
+        self._public_key = public_key
+        self._route = route
+
+        @web.middleware
+        async def interaction_endpoint_middleware(request: web.Request, handler):
+            print("MIDDLEWARE TRIGGERED.")
+            if request.path == self._route and request.method == "POST":
+                logger.debug("Request POSTing on interaction endpoint route, intercepting.")
+                return await self._on_interaction_incoming(request)
+            else:
+                resp = await handler(request)
+                return resp
+
+        return interaction_endpoint_middleware
+
+    async def start(self, public_key: str, *, port: int = 8080, route="/interactions"):
+        app = web.Application(middlewares=[self.middleware(public_key, route)])
+        runner = web.AppRunner(app)
+        await runner.setup()
+        site = web.TCPSite(runner, "0.0.0.0", port)
+        self.event_dispatcher.add_listener(self._handle_ping, "ping")
+        await site.start()
+        logger.info("%s listening on %s", self.__class__.__name__, site.name)
+
+    def run(
+            self,
+            public_key: str,
+            *,
+            loop: asyncio.AbstractEventLoop | None = None,
+            port: int = 8080,
+            route: str = "/interactions",
+    ):
+        loop = loop or asyncio.new_event_loop()
+        task = loop.create_task(self.start(public_key, port=port, route=route))
+        try:
+            loop.run_forever()
+        except KeyboardInterrupt:
+            logger.debug("KeyboardInterrupt encountered, stopping loop.")
+            task.cancel()
```

Comment:
I think a clean tasks function, like the one from the asyncio runner, should be used here considering the dispatcher currently leaves tasks in the event loop instead of cleaning them up.